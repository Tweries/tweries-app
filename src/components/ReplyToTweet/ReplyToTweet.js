import classnames from 'classnames';
import React, { useEffect, useState } from 'react';
import fetchTweet from '../../api/fetchTweet';

const copy = {
  'Optional: reply to Tweet URL': 'Optional: reply to Tweet URL'
};

// TODO: move to containers
function ReplyToTweet({ callback, onChange, tweetUrl, userId }) {
  const [isValidTweet, setIsValidTweet] = useState(false);
  const [waiting, setWaiting] = useState(false);

  useEffect(() => {
    async function fetchData() {
      try {
        const response = await fetchTweet({
          inReplyToTweetUrl: tweetUrl,
          userId
        });
        console.log(response);
        if (response.data) {
          callback(null, response.data);
          setIsValidTweet(true);
          setWaiting(false);
        } else {
          callback(response.error);
          setIsValidTweet(false);
          setWaiting(false);
        }
      } catch (error) {
        callback(error);
        setIsValidTweet(false);
        setWaiting(false);
      }
    }
    if (tweetUrl && userId) {
      setWaiting(true);
      fetchData();
    }
  }, [callback, tweetUrl, userId]);

  return (
    <>
      <label className="text-sm" htmlFor="reply-to">
        {copy['Optional: reply to Tweet URL']}
      </label>
      <textarea
        className={classnames('p-2 tweries-border', {
          'tweries-background-color-blue-white':
            waiting === true || tweetUrl === '',
          'bg-green-200': waiting === false && tweetUrl !== '' && isValidTweet,
          'bg-red-200': waiting === false && tweetUrl !== '' && !isValidTweet
        })}
        data-testid="reply-to"
        name="reply-to"
        onChange={({ target: { value } }) => {
          onChange(value);
        }}
        placeholder={copy['Optional: reply to Tweet URL']}
        rows={1}
        value={tweetUrl}
      />
    </>
  );
}

export default ReplyToTweet;
