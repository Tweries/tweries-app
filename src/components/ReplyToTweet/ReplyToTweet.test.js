import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import ReplyToTweet from './ReplyToTweet';

const mockOnChangeId = jest.fn();
const mockOnChangeUrl = jest.fn();

const baseProps = {
  onChangeId: mockOnChangeId,
  onChangeUrl: mockOnChangeUrl,
  userId: null,
  value: ''
};

const TWEET_URL = 'https://twitter.com/musk_china/status/1199474666412236800';
const USER_ID = 'twitter|1183836409850814464';

describe('ReplyToTweet', () => {
  beforeEach(() => {
    mockOnChangeId.mockReset();
    mockOnChangeUrl.mockReset();
  });

  it('to match snapshot', () => {
    const props = { ...baseProps };
    const { container } = render(<ReplyToTweet {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('good URL', () => {
    fetch.mockResponseOnce(JSON.stringify({ statusId: '1199474666412236800' }));

    const props = {
      ...baseProps,
      userId: USER_ID,
      value: TWEET_URL
    };
    render(<ReplyToTweet {...props} />);

    // TODO: add assertion
  });

  it('bad URL', () => {
    fetch.mockResponseOnce(JSON.stringify({}));

    const props = {
      ...baseProps,
      userId: USER_ID,
      value: 'something'
    };
    render(<ReplyToTweet {...props} />);

    // TODO: add assertion
  });

  it('change URL', () => {
    fetch.mockResponseOnce(JSON.stringify({}));

    const props = {
      ...baseProps,
      userId: USER_ID,
      value: ''
    };
    const { getByTestId } = render(<ReplyToTweet {...props} />);

    fireEvent.change(getByTestId('reply-to'), {
      target: { value: TWEET_URL }
    });

    // TODO: add assertion
  });
});
