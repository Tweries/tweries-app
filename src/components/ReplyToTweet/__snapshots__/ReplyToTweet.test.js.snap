// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReplyToTweet to match snapshot 1`] = `
<div>
  <label
    class="text-sm"
    for="reply-to"
  >
    Optional: reply to Tweet URL
  </label>
  <textarea
    class="p-2 tweries-border tweries-background-color-blue-white"
    data-testid="reply-to"
    name="reply-to"
    placeholder="Optional: reply to Tweet URL"
    rows="1"
  />
</div>
`;
